src/lib/barcode_matching.rs

```diff
@@ -108,16 +109,21 @@ impl BarcodeMatcher {
         let mut best_barcode_index = self.sample_barcodes.len();
         let mut best_mismatches = 255u8;
         let mut next_best_mismatches = 255u8;
+        let mut max_mismatches = 255u8;
+        let read_bases = encode(read_bases); // NB: this encodes IUPAC bases in the read, but count_mismatches will treat them as no-calls.
         for (index, sample_barcode) in self.sample_barcodes.iter().enumerate() {
-            let mismatches = Self::count_mismatches(read_bases, sample_barcode.as_bstr());
-
+            let mismatches = Self::count_mismatches(&read_bases, sample_barcode, max_mismatches);
             if mismatches < best_mismatches {
                 next_best_mismatches = best_mismatches;
                 best_mismatches = mismatches;
                 best_barcode_index = index;
             } else if mismatches < next_best_mismatches {
                 next_best_mismatches = mismatches;
             }
+            if next_best_mismatches < 255u8 - self.min_mismatch_delta {
```

Comment:
I don't think you need to check this at all if you don't update `next_best_mismatches`, right?

```
if mismatches < next_best_mismatches {
    if mismatches < best_mismatches {
        next_best_mismatches = best_mismatches;
        best_mismatches = mismatches;
        best_barcode_index = index;
    } else {
        next_best_mismatches = mismatches;
    }
    if next_best_mismatches < 255u8 - self.min_mismatch_delta {
        max_mismatches =
            min(max_mismatches, next_best_mismatches + self.min_mismatch_delta);
    }
}

Reply:
The change from `else if ` to `else` in your suggestion may give the wrong value to `next_best_mismatches`, since it is always updated, even if `mismatches < next_best_mismatches`. 

But I see your point in that we don't need new check at all if we don't update `next_best_mismatches`.  But would we duplicate this in the `if` and `else if`?  I'll make the update